README.md
## Convolutional Neural Networks for Sentence Classification

Fork by Giuseppe Attardi from https://github.com/yoonkim/CNN_sentence.git.
Differences with the original:
- multiclass classification
- model saving and loading for classifying files, e.g. tweets
- configurable parameters
- integrated program, without a separate preprocessing stage
- input in a single file, annotated in the format used in the SemEval Twitter Sentiment Analysis tasks.

The original code by Kim Yoon implements the technique described in the paper:
[Convolutional Neural Networks for Sentence Classification](http://arxiv.org/abs/1408.5882) (EMNLP 2014).
Please cite both this page and the original paper when using the program.

### Requirements
Code is written in Python (2.7) and requires Theano (0.7).

Using the pre-trained `word2vec` vectors will also require downloading the binary file from
https://code.google.com/p/word2vec/


### Invocation
```
usage: conv_net_sentence.py [-h] [-train] [-static] [-vectors VECTORS] [-filters FILTERS]
                            [-clean] [-dropout DROPOUT] [-epochs EPOCHS]
                            [-tagField TAGFIELD] [-textField TEXTFIELD]
                            model input

CNN sentence classifier.

positional arguments:
  model                 model file (default mr)
  input                 train/test file in SemEval twitter format

optional arguments:
  -h, --help            show this help message and exit
  -train                train model
  -static               static or nonstatic
  -clean                tokenize text
  -filters FILTERS      n[,n]* (default 3,4,5)
  -vectors VECTORS      word2vec embeddings file (random values if missing)
  -dropout DROPOUT      dropout probability (default 0.5)
  -epochs EPOCHS        training iterations (default 25)
  -tagField TAGFIELD    label field in files (default 1)
  -textField TEXTFIELD  text field in files (default 2)
```

VECTORS is the word2vec binary file (e.g. `GoogleNews-vectors-negative300.bin` file),
clean, if present, text is lower-cased and tokenized,

input contain sentences from the corpus in SemEval format, i.e. one sentence
per line, tab separated values:

ID	UID	text

Note: This will create the dataset with different fold-assignments than was used in the paper.
You should still be getting a CV score of >81% with CNN-nonstatic model, though.

### Running the models (CPU)
Example commands:

```
THEANO_FLAGS=mode=FAST_RUN,device=cpu,floatX=float32 ./conv_net_sentence.py mr.rand mr.tsv
THEANO_FLAGS=mode=FAST_RUN,device=cpu,floatX=float32 ./conv_net_sentence.py -vectors w2v_file -static mr.w2v-static mr.tsv
THEANO_FLAGS=mode=FAST_RUN,device=cpu,floatX=float32 ./conv_net_sentence.py -vectors w2v_file mr.w2v mr.tsv
```

This will run the CNN-rand, CNN-static, and CNN-nonstatic models respectively in the paper.

### Using the GPU
GPU will result in a good 10x to 20x speed-up, so it is highly recommended. 
To use the GPU, simply change `device=cpu` to `device=gpu` (or whichever gpu you are using).
For example:
```
THEANO_FLAGS=mode=FAST_RUN,device=gpu,floatX=float32 python conv_net_sentence.py -nonstatic -word2vec
```

### Example output
CPU output:
```
epoch: 1, training time: 219.72 secs, train perf: 81.79 %, val perf: 79.26 %
epoch: 2, training time: 219.55 secs, train perf: 82.64 %, val perf: 76.84 %
epoch: 3, training time: 219.54 secs, train perf: 92.06 %, val perf: 80.95 %
```
GPU output:
```
epoch: 1, training time: 16.49 secs, train perf: 81.80 %, val perf: 78.32 %
epoch: 2, training time: 16.12 secs, train perf: 82.53 %, val perf: 76.74 %
epoch: 3, training time: 16.16 secs, train perf: 91.87 %, val perf: 81.37 %
```

### Other Implementations
#### TensorFlow
[Denny Britz](http://www.wildml.com) has an implementation of the model in TensorFlow:

https://github.com/dennybritz/cnn-text-classification-tf

He also wrote a [nice tutorial](http://www.wildml.com/2015/12/implementing-a-cnn-for-text-classification-in-tensorflow) on it, as well as a general tutorial on [CNNs for NLP](http://www.wildml.com/2015/11/understanding-convolutional-neural-networks-for-nlp).

### Hyperparameters
At the time of my original experiments I did not have access to a GPU so I could not run a lot of different experiments.
Hence the paper is missing a lot of things like ablation studies and variance in performance.

Ye Zhang has written a [very nice paper](http://arxiv.org/abs/1510.03820) doing an extensive analysis of model variants (e.g. filter widths, k-max pooling, word2vec vs Glove, etc.) and their effect on performance.

### SemEval-2015 Task 10: Sentiment Analysis in Twitter

Experiments using the data from SemEval-15, using option --filters 7,7,7, achieves an accuracy
of 67.28%, better than the top official score (64.84%), as reported in:
Sara Rosenthal et al. [SemEval-2015 Task 10: Sentiment Analysis in Twitter](http://alt.qcri.org/semeval2015/cdrom/pdf/SemEval078.pdf).



ChangeLog
2016-04-17  Giuseppe Attardi  <attardi@R730-gpu>

	* updates.py: various learning rules, including AdaGrad, AdaDelta, Adam.
	Also includes regularization methods.

	* layers.py: more modular version of classes: each layer provides method output() to
	generate predictions.

2016-03-28  Giuseppe Attardi  <attardi@R730-gpu>

	* conv_net_classes.py: updated deprecated theano.tensor.signal.downsample.conv2d with
	theano.tensor.nnet.conv2d

2016-03-27  Giuseppe Attardi  <attardi@R730-gpu>

	* conv_net_classes.py (ConvNet.save): save/load self.activations

2016-02-26  Giuseppe Attardi  <attardi@R730-gpu>

	* process_data.py (load_word2vec): load vectors also in text format.



process_data.py

import numpy as np
from collections import defaultdict


def load_sentences(train_file, tagField=1, textField=2):
    """
    Loads sentences.
    :param train_file: filename containing labeled sentences in TSV format.
    :return: sents (paired with labels), word doc freq, list of labels.
    """
    sents = []
    tags = {}
    word_df = defaultdict(int)
    with open(train_file, "rb") as f:
        for line in f:       
            fields = line.strip().split("\t")
            text = fields[textField]
            tag = fields[tagField]
            if tag not in tags:
                tags[tag] = len(tags)
            clean_text = text.lower()
            words = clean_text.split()
            for word in set(words):
                word_df[word] += 1
            pair = (words, tags[tag])
            sents.append(pair)
    labels = [0] * len(tags)
    for tag,i in tags.iteritems():
        labels[i] = tag
    return sents, word_df, labels


def load_vectors(fname, binary=True):
    """
    Loads word vectors from file in word2vec format.
    :param fname: name of file in word2vec format.
    :return: vectors and word list.
    """
    with open(fname, "rb") as f:
        header = f.readline()
        vocab_size, embeddings_size = map(int, header.split())
        words = [''] * vocab_size
        vectors = np.empty((vocab_size, embeddings_size), dtype='float32')
        if binary:
            binary_len = np.dtype('float32').itemsize * embeddings_size
            for i in xrange(vocab_size):
                word = []
                while True:
                    ch = f.read(1)
                    if ch == ' ':
                        word = ''.join(word)
                        break
                    if ch != '\n':
                        word.append(ch)   
                words[i] = word
                vectors[i,:] = np.fromstring(f.read(binary_len), dtype='float32')  
        else:                   # text
            for i,line in enumerate(f):
                items = line.split()
                words[i] = unicode(items[0], 'utf-8')
                vectors[i,:] = np.array(map(float, items[1:]))
    return vectors, words


def load_word_vectors(fname, word_index, binary=True):
    """
    Loads word vectors from file in word2vec format.
    :param fname: name of file in word2vec format.
    :return: vectors and word list.
    """
    with open(fname, "rb") as f:
        header = f.readline()
        vocab_size, embeddings_size = map(int, header.split())
        vectors = np.zeros((len(word_index), embeddings_size), dtype='float32')
        if binary:
            binary_len = np.dtype('float32').itemsize * embeddings_size
            for i in xrange(vocab_size):
                word = []
                while True:
                    ch = f.read(1)
                    if ch == ' ':
                        word = ''.join(word)
                        break
                    if ch != '\n':
                        word.append(ch)   
                if word in word_index:
                    vectors[word_index[word],:] = np.fromstring(f.read(binary_len), dtype='float32')  
                else:
                    f.read(binary_len)
        else:                   # text
            for i,line in enumerate(f):
                items = line.split()
                word = unicode(items[0], 'utf-8')
                if word in word_index:
                    vectors[word_index[word],:] = np.array(map(float, items[1:]))
    high = 2.38 / np.sqrt(len(vectors) + embeddings_size) # see (Bottou '88)
    for i,v in enumerate(vectors):
        if np.count_nonzero(v) == 0:
            vectors[i:] = np.random.uniform(-high, high, embeddings_size)
    return np.asarray(vectors, dtype="float32")

def add_unknown_words(vectors, words, word_df, k, min_df=1):
    """
    Create word vector for words in :param word_df: that occur in at least :param min_df: documents.
    :param word_df: dictionary of word document frequencies.
    :param k: size of embedding vectors.
    """
    wordset = set(words)
    high = 2.38 / np.sqrt(len(vectors)) # see (Bottou '88)
    start = len(vectors)
    end = start
    for word in word_df:
        if word not in wordset and word_df[word] >= min_df:
            end += 1
            words.append(word)
    vectors.resize((end, k), refcheck=False)
    for i in range(start, end):
        vectors[i:] = np.random.uniform(-high, high, k)


conv_net_tweets.py
#!/usr/bin/env python

"""
Training a convolutional network for sentence classification,
as described in paper:
Convolutional Neural Networks for Sentence Classification
http://arxiv.org/pdf/1408.5882v2.pdf
"""
import cPickle as pickle
import numpy as np
import theano
import sys
import argparse
import warnings
warnings.filterwarnings("ignore")   

# run from everywhere without installing
sys.path.append(".")
from conv_net_classes import *
from process_data import process_data


def sent2indices(sent, word_index, max_l, pad):
    """
    Transforms sentence into a list of indices. Pad with zeroes.
    Drop words non in word_index.
    :param sent: list of words.
    :param word_index: associates an index to each word
    :param max_l: max sentence length
    :param pad: pad size
    """
    x = [0] * pad                # left padding
    for word in sent:
        if word in word_index: # FIXME: skips unknown words
            if len(x) < max_l: # truncate long sent
                x.append(word_index[word])
            else:
                break
    # len(x) includes pad
    rpad = [0] * max(0, max_l + 2 * pad - len(x)) # right padding
    return x + rpad


def read_corpus(filename, word_index, max_l, pad=2, clean_string=False,
                textField=3):
    """
    Load test corpus, in TSV format.
    :param filename: file with sentences.
    :param word_index: word IDs.
    :param max_l: max sentence length.
    :param pad: padding size.
    :param textField: index of field containing text.
    :return: an array, each row consists of sentence word indices
    """
    corpus = []
    with open(filename) as f:
        for line in f:
            fields = line.strip().split("\t")
            text = fields[textField]
            if clean_string:
                text_clean = clean_str(text)
            else:
                text_clean = text.lower()
            # turn sentences into lists of indices
            sent = sent2indices(text_clean.split(), word_index, max_l, pad)
            corpus.append(sent)
    return np.array(corpus, dtype="int32")


def predict(cnn, test_set_x):

    test_set_y_pred = cnn.predict(test_set_x)
    # compile expression
    test_function = theano.function([cnn.x], test_set_y_pred, allow_input_downcast=True)
    return test_function(test_set_x)


if __name__=="__main__":
    parser = argparse.ArgumentParser(description="CNN sentence classifier.")
    
    parser.add_argument('model', type=str, default='mr',
                        help='model file (default %(default)s)')
    parser.add_argument('input', type=str,
                        help='train/test file in SemEval twitter format')
    parser.add_argument('-train', help='train model',
                        action='store_true')
    parser.add_argument('-clean', help='tokenize text',
                        action='store_true')
    parser.add_argument('-filters', type=str, default='3,4,5',
                        help='n[,n]* (default %(default)s)')
    parser.add_argument('-vectors', type=str,
                        help='word2vec embeddings file (random values if missing)')
    parser.add_argument('-dropout', type=float, default=0.5,
                        help='dropout probability (default %(default)s)')
    parser.add_argument('-hidden', type=int, default=100,
                        help='hidden units in feature map (default %(default)s)')
    parser.add_argument('-epochs', type=int, default=100,
                        help='training iterations (default %(default)s)')
    parser.add_argument('-tagField', type=int, default=1,
                        help='label field in files (default %(default)s)')
    parser.add_argument('-textField', type=int, default=2,
                        help='text field in files (default %(default)s)')

    args = parser.parse_args()

    # theano.config
    theano.config.floatX = 'float32'

    if not args.train:
        # predict
        with open(args.model) as mfile:
            cnn = ConvNet.load(mfile)
            word_index, max_l, pad, labels = pickle.load(mfile)
        test_set_x = read_corpus(args.input, word_index, max_l, pad, textField=args.textField)
        results = predict(cnn, test_set_x)
        # convert indices to labels
        for line, y in zip(open(args.input), results):
            tokens = line.split("\t")
            tokens[args.tagField] = labels[y]
            print "\t".join(tokens),
        sys.exit()

    # training
    sents, U, word_index, vocab, labels = process_data(args.input, args.clean,
                                                       args.vectors,
                                                       args.tagField,
                                                       args.textField)

    # sents is a list of pairs: (list of words, label)
    # vocab: dict of word doc freq
    filter_hs = [int(x) for x in args.filters.split(',')]
    model = args.model
    if args.vectors:
        print "using: word2vec vectors"
    else:
        print "using: random vectors"

    # filter_h determines padding, hence it depends on largest filter size.
    pad = max(filter_hs) - 1
    max_l = max(len(x_y[0]) for x_y in sents)
    height = max_l + 2 * pad    # padding on both sides
    width = U.shape[1]
    feature_maps = args.hidden
    output_units = len(labels)
    batch_size = 50
    conv_activation = "relu"
    activation = Iden #T.tanh         # Iden
    dropout_rate = args.dropout
    sqr_norm_lim = 9
    shuffle_batch = True
    lr_decay = 0.95
    parameters = (("image shape", height, width),
                  ("filters", args.filters),
                  ("feature maps", feature_maps),
                  ("output units", output_units),
                  ("dropout rate", dropout_rate),
                  ("batch size", batch_size),
                  ("adadelta decay", lr_decay),
                  ("conv_activation", conv_activation),
                  ("activation", activation),
                  ("sqr_norm_lim", sqr_norm_lim),
                  ("shuffle batch", shuffle_batch))
    for param in parameters:
        print "%s: %s" % (param[0], ",".join(str(x) for x in param[1:]))

    cnn = ConvNet(U, height, width,
                  filter_hs=filter_hs,
                  conv_activation=conv_activation,
                  feature_maps=feature_maps,
                  output_units=output_units,
                  batch_size=batch_size,
                  dropout_rates=[dropout_rate],
                  activations=[activation])
    # each item in train is a list of indices for each sentencs plus the id of the label
    train = [sent2indices(words, word_index, max_l, pad) + [y]
             for words,y in sents]
    train_set = np.array(train, dtype="int32")

    # model saver
    def save():
        with open(model, "wb") as mfile:
            cnn.save(mfile)
            pickle.dump((word_index, max_l, pad, labels), mfile)

    cnn.train(train_set, epochs=args.epochs,
              lr_decay=lr_decay,
              shuffle_batch=shuffle_batch, 
              sqr_norm_lim=sqr_norm_lim,
              save=save)


conv_net_sentence.py
#!/usr/bin/env python

"""
Training a convolutional network for sentence classification,
as described in paper:
Convolutional Neural Networks for Sentence Classification
http://arxiv.org/pdf/1408.5882v2.pdf
"""
import cPickle as pickle
import numpy as np
import theano
import sys
import argparse
import warnings
warnings.filterwarnings("ignore")   

# run from everywhere without installing
sys.path.append(".")
from conv_net_classes import *
from process_data import process_data


def get_idx_from_sent(sent, word_index, max_l, pad):
    """
    Transforms sentence into a list of indices. Pad with zeroes.
    Drop words non in word_index.
    :param sent: list of words.
    :param word_index: associates an index to each word
    :param max_l: max sentence length
    :param pad: pad length
    """
    x = [0] * pad                # left padding
    for word in sent:
        if word in word_index: # FIXME: skips unknown words
            if len(x) < max_l: # truncate long sent
                x.append(word_index[word])
            else:
                break
    # len(x) includes pad
    rpad = [0] * max(0, max_l + 2 * pad - len(x)) # right padding
    return x + rpad


def make_idx_data_cv(revs, word_index, cv, max_l, pad):
    """
    Transforms sentences into a 2-d matrix and splits them into
    train and test according to cv.
    :param revs: corpus partitioned into folds
    :param word_index: maps word to IDs
    :param cv: cross-validation step, -1 for no cross-validation
    :param max_l: max sentence length
    :param pad: pad length
    """
    train, test = [], []
    for rev in revs:
        sent = get_idx_from_sent(rev["text"], word_index, max_l, pad)
        sent.append(rev["y"])
        if rev["split"] == cv:
            test.append(sent)        
        else:  
            train.append(sent)   
    train = np.array(train, dtype="int32")
    test = np.array(test, dtype="int32")
    return train, test
  

def read_corpus(filename, word_index, max_l, pad=2, clean_string=False,
                textField=3):
    """
    Load test corpus, in TSV format.
    :param filename: file with sentences.
    :param word_index: word IDs.
    :param max_l: max sentence length.
    :param pad: padding size.
    :param textField: index of field containing text.
    :return: an array, each row consists of sentence word indices
    """
    corpus = []
    with open(filename) as f:
        for line in f:
            fields = line.strip().split("\t")
            text = fields[textField]
            if clean_string:
                text_clean = clean_str(text)
            else:
                text_clean = text.lower()
            sent = get_idx_from_sent(text_clean, word_index, max_l, pad)
            corpus.append(sent)
    return np.array(corpus, dtype="int32")


def predict(cnn, test_set_x):

    test_set_y_pred = cnn.predict(test_set_x)
    # compile expression
    test_function = theano.function([cnn.x], test_set_y_pred, allow_input_downcast=True)
    return test_function(test_set_x)


if __name__=="__main__":
    parser = argparse.ArgumentParser(description="CNN sentence classifier.")
    
    parser.add_argument('model', type=str, default='mr',
                        help='model file (default %(default)s)')
    parser.add_argument('input', type=str,
                        help='train/test file in SemEval twitter format')
    parser.add_argument('-train', help='train model',
                        action='store_true')
    parser.add_argument('-clean', help='tokenize text',
                        action='store_true')
    parser.add_argument('-cv', type=int, default=10,
                        help='perform cross validation, 0 for no cross varlidation')
    parser.add_argument('-filters', type=str, default='3,4,5',
                        help='n[,n]* (default %(default)s)')
    parser.add_argument('-vectors', type=str,
                        help='word2vec embeddings file (random values if missing)')
    parser.add_argument('-dropout', type=float, default=0.5,
                        help='dropout probability (default %(default)s)')
    parser.add_argument('-hidden', type=int, default=100,
                        help='hidden units in feature map (default %(default)s)')
    parser.add_argument('-epochs', type=int, default=25,
                        help='training iterations (default %(default)s)')
    parser.add_argument('-tagField', type=int, default=1,
                        help='label field in files (default %(default)s)')
    parser.add_argument('-textField', type=int, default=2,
                        help='text field in files (default %(default)s)')

    args = parser.parse_args()

    # theano.config
    theano.config.floatX = 'float32'

    if not args.train:
        # predict
        with open(args.model) as mfile:
            cnn = ConvNet.load(mfile)
            word_index, max_l, pad, labels = pickle.load(mfile)

        test_set_x = read_corpus(args.input, word_index, max_l, pad, textField=args.textField)
        results = predict(cnn, test_set_x)
        # convert indices to labels
        for line, y in zip(open(args.input), results):
            tokens = line.split("\t")
            tokens[args.tagField] = labels[y]
            print "\t".join(tokens),
        sys.exit()

    # training
    sents, U, word_index, vocab, labels = process_data_cv(args.input, args.clean,
                                                          args.vectors,
                                                          args.tagField,
                                                          args.textField,
                                                          args.cv)

    # sents is a list of entries, where each entry is a dict:
    # {"y": 0/1, "text": , "num_words": , "split": cv fold}
    # vocab: dict of word doc freq
    filter_hs = [int(x) for x in args.filters.split(',')]
    model = args.model
    if args.vectors:
        print "using: word2vec vectors"
    else:
        print "using: random vectors"

    # filter_h determines padding, hence it depends on largest filter size.
    pad = max(filter_hs) - 1
    max_l = max(len(s["num_words"]) for s in sents)
    height = max_l + 2 * pad # padding on both sides

    classes = set(x["y"] for x in sents)
    width = U.shape[1]
    conv_non_linear = "relu"
    feature_maps = args.hidden
    output_units = len(labels)
    batch_size = 50
    dropout_rate = args.dropout
    sqr_norm_lim = 9
    shuffle_batch = True
    lr_decay = 0.95
    parameters = (("image shape", height, width),
                  ("filters", args.filters),
                  ("feature maps units", feature_maps),
                  ("dropout rate", dropout_rate),
                  ("batch size", batch_size),
                  ("adadelta decay", lr_decay),
                  ("conv_non_linear", conv_non_linear),
                  ("non static", non_static),
                  ("sqr_norm_lim", sqr_norm_lim),
                  ("shuffle batch", shuffle_batch))
    for param in parameters:
        print "%s: %s" % (param[0], ",".join(str(x) for x in param[1:]))
    results = []
    # ensure replicability
    np.random.seed(3435)
    # perform n-fold cross-validation
    for i in range(0, (1, args.cv)):
        # test = [padded(x) for x in sents if x[split] == i]
        # train is rest
        if args.cv < 2:
            cv = -1             # no cross validation
        else:
            cv = args.cv
        train_set, test_set = make_idx_data_cv(sents, word_index, i, max_l, pad)
        cnn = ConvNet(U, height, width,
                      filter_hs=filter_hs,
                      conv_activation=conv_non_linear,
                      feature_maps=feature_maps,
                      output_units=output_units,
                      batch_size=batch_size,
                      dropout_rates=[dropout_rate],
                      activations=[activation])
        perf = cnn.train(train_set,
                         lr_decay=lr_decay,
                         shuffle_batch=shuffle_batch, 
                         epochs=args.epochs,
                         sqr_norm_lim=sqr_norm_lim)
        print "cv: %d, perf: %f" % (i, perf)
        # save model
        if i == 0 or perf > max(results):
            with open(model, "wb") as mfile:
                cnn.save(mfile)
                pickle.dump((word_index, max_l, pad, labels), mfile)
        results.append(perf)  
    print "Avg. accuracy: %.4f" % np.mean(results)
